From 4eb18dd487da4cb621dcbecfc475fc0871b356ac Mon Sep 17 00:00:00 2001
From: Ma Jianpeng <jianpeng.ma@intel.com>
Date: Wed, 23 Jul 2014 10:10:38 -0700
Subject: [PATCH] os/FileJournal: Update the journal header when closing
 journal

When closing journal, it should check must_write_header and update
journal header if must_write_header alreay set.
It can reduce the nosense journal-replay after restarting osd.

Signed-off-by: Ma Jianpeng <jianpeng.ma@intel.com>
Reviewed-by: Sage Weil <sage@redhat.com>
---
 src/os/FileJournal.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/src/os/FileJournal.cc
+++ b/src/os/FileJournal.cc
@@ -543,8 +543,9 @@
   stop_writer();
 
   // close
   assert(writeq_empty());
+  assert(!must_write_header);
   assert(fd >= 0);
   VOID_TEMP_FAILURE_RETRY(::close(fd));
   fd = -1;
 }
@@ -1101,9 +1102,9 @@
   dout(10) << "write_thread_entry start" << dendl;
   while (1) {
     {
       Mutex::Locker locker(writeq_lock);
-      if (writeq.empty()) {
+      if (writeq.empty() && !must_write_header) {
 	if (write_stop)
 	  break;
 	dout(20) << "write_thread_entry going to sleep" << dendl;
 	writeq_cond.Wait(writeq_lock);
